compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/BallerinaSemanticModel.java

```diff
@@ -183,40 +183,84 @@ public BallerinaSemanticModel(BLangPackage bLangPackage, CompilerContext context
+        Location symbolLocation = symbolAtCursor.getPosition();
+        BLangNode node = findEnclosingContainerNode(symbolLocation.lineRange(), null);
 
-        BLangNode node = new NodeFinder(false)
-                .lookupEnclosingContainer(this.bLangPackage, symbolLocation.get().lineRange());
+        return getReferences(symbolAtCursor, node, withDefinition);
+    }
 
-        ReferenceFinder refFinder = new ReferenceFinder(withDefinition);
-        return refFinder.findReferences(node, getInternalSymbol(symbol));
+    @Override
+    public List<Location> references(Symbol symbol, Document targetDocument, boolean withDefinition) {
+        BSymbol symbolAtCursor = getInternalSymbol(symbol);
+        Optional<Location> symbolLocation = symbol.getLocation();
+        if (symbolAtCursor == null || symbolLocation.isEmpty()) {
+            return Collections.emptyList();
+        }
+        BLangNode node = findEnclosingContainerNode(symbolLocation.get().lineRange(), targetDocument);
+
+        return getReferences(symbolAtCursor, node, withDefinition);
     }
 
     @Override
-    public List<Location> references(Document sourceDocument, LinePosition position, boolean withDefinition) {
-        BLangCompilationUnit compilationUnit = getCompilationUnit(sourceDocument);
-        SymbolFinder symbolFinder = new SymbolFinder();
-        BSymbol symbolAtCursor = symbolFinder.lookup(compilationUnit, position);
+    public List<Location> references(Document sourceDocument,
+                                     Document targetDocument,
+                                     LinePosition position,
+                                     boolean withDefinition) {
 
+        BSymbol symbolAtCursor = findSymbolAtCursorPosition(sourceDocument, position);
         if (symbolAtCursor == null) {
             return Collections.emptyList();
         }
+        Location symbolLocation = symbolAtCursor.getPosition();
+        BLangNode node = findEnclosingContainerNode(symbolLocation.lineRange(), targetDocument);
 
-        BLangNode node = new NodeFinder(false)
-                .lookupEnclosingContainer(this.bLangPackage, symbolAtCursor.pos.lineRange());
+        return getReferences(symbolAtCursor, node, withDefinition);
+    }
+
+    private BSymbol findSymbolAtCursorPosition(Document sourceDocument, LinePosition linePosition) {
+        BLangCompilationUnit sourceCompilationUnit = getCompilationUnit(sourceDocument);
+        SymbolFinder symbolFinder = new SymbolFinder();
+        return symbolFinder.lookup(sourceCompilationUnit, linePosition);
+    }
+
+    private BLangNode findEnclosingContainerNode(LineRange lineRange, Document targetDocument) {
+        return targetDocument == null ?
+                new NodeFinder(false)
+                        .lookupEnclosingContainer(this.bLangPackage, lineRange) :
+                new NodeFinder(false)
+                        .lookupEnclosingContainer(getCompilationUnit(targetDocument), lineRange);
+    }
```

Comment:
This method is not required right? This is basically a one-liner, the case to be used depends on the particular `references()` version